src/blockrelay/graphene.cpp

```diff
@@ -1334,6 +1413,191 @@ bool HandleGrapheneBlockRequest(CDataStream &vRecv, CNode *pfrom, const CChainPa
     return true;
 }
 
+bool HandleGrapheneBlockRecoveryRequest(CDataStream &vRecv, CNode *pfrom, const CChainParams &chainparams)
+{
+    CRequestGrapheneReceiverRecover recoveryRequest;
+    vRecv >> recoveryRequest;
+
+    std::shared_ptr<CGrapheneBlock> grapheneBlock = thinrelay.GetSentGrapheneBlocks(pfrom->GetId());
+    if (!grapheneBlock)
+        return error("No block available to reconstruct for get_grrec");
+
+    // We had a block stored but it was the wrong one
+    if (grapheneBlock->header.GetHash() != recoveryRequest.blockhash)
+        return error("Sender does not have block for requested hash");
+
+    CGrapheneReceiverRecover recoveryResponse = CGrapheneReceiverRecover(
+        *recoveryRequest.pReceiverFilter, *grapheneBlock, recoveryRequest.nSenderFilterPositives, pfrom);
+    pfrom->PushMessage(NetMsgType::GRAPHENE_RECOVERY, recoveryResponse);
+
+    return true;
+}
+
+bool HandleGrapheneBlockRecoveryResponse(CDataStream &vRecv, CNode *pfrom, const CChainParams &chainparams)
+{
+    CGrapheneReceiverRecover recoveryResponse;
+    vRecv >> recoveryResponse;
+
+    auto pblock = thinrelay.GetBlockToReconstruct(pfrom, recoveryResponse.blockhash);
+    if (pblock == nullptr)
+        return error("No block available to reconstruct for grrec");
+    DbgAssert(pblock->grapheneblock != nullptr, return false);
+    CGrapheneBlock grapheneBlock = *(pblock->grapheneblock);
+
+    CIblt localIblt((*recoveryResponse.pRevisedIblt));
+    localIblt.reset();
+
+    // Initialize map with txs from various pools
+    std::map<uint64_t, uint256> mapTxFromPools;
+    pblock->grapheneblock->FillTxMapFromPools(mapTxFromPools);
+
+    // Insert additional txs and identify coinbase
+    CTransactionRef coinbase = nullptr;
+    for (auto &tx : pblock->grapheneblock->vAdditionalTxs)
+    {
+        const uint256 &hash = tx->GetHash();
+        uint64_t cheapHash = pblock->grapheneblock->pGrapheneSet->GetShortID(hash);
+
+        mapTxFromPools.insert(std::make_pair(cheapHash, hash));
+
+        if (tx->IsCoinBase())
+            coinbase = tx;
+    }
+
+    if (coinbase == nullptr)
+    {
+        LOG(GRAPHENE, "Error: No coinbase transaction found in graphene block, peer=%s", pfrom->GetLogName());
+        return false;
+    }
+
+    // Insert latest transactions just sent over
+    for (auto &tx : recoveryResponse.vMissingTxs)
+    {
+        uint256 hash = tx.GetHash();
+        uint64_t cheapHash = pblock->grapheneblock->pGrapheneSet->GetShortID(hash);
```

Comment:
small nit, you can take a reference here .

-uint256 hash = tx.GetHash()
+const uint256 &hash = tx.GetHash()

Reply:
Done.